### DISCUSSION TITLE
CompositionRoot: GetViewFor

### DISCUSSION THREAD

**Discussion by kgday** [JordanMarr/ReactiveElmish.Avalonia#52] [Ideas]
If we had additional methods to get a view for an already existing viewmodel, we could replace the existing view location to get the view from the container for a given view model rather than create through reflection.

I imagine the new methods to support this could be implemented something like this:
```
    ///Check to see if a view has been registered for the given view model
    member this.HasViewFor(vm: 'ViewModel & #ReactiveElmishViewModel) =
        let vmType = vm.GetType()
        viewRegistty.Value |> Map.tryFind(VM.Key vmType) |> Option.isSome

    ///Gets or creates a view for a given view model instance and binds it
    member this.GetViewFor(vm: 'ViewModel & #ReactiveElmishViewModel) =
        let vmType = vm.GetType()
        match viewRegistry.Value |> Map.tryFind (VM.Key vmType) with
        | Some registration -> 
            match registration with
            | SingletonView view ->
                if view.DataContext = null then
                    ViewBinder.bind (vm, view) |> snd
                else
                    view
            | TransientView createView -> ViewBinder.bindWithDisposeOnViewUnload (vm, createView()) |> snd
        | None ->
            failwithf $"No view registered for VM type {vmType.FullName}"
```

That way I can add the following to my application:
module Root =
    let lazyInstance= lazy AppCompositionRoot()
    let public current = lazyInstance.Value

In my App,axaml.fs 
instead of
`let appRoot = MyAppComppsite()`

I would have 
`let appRoot = Root.current`

And then my ViewLocator() class implementing IDataTemplate could also call Root.current to get the view for a given view model so that it can be statically instantiated in the zaml is it is currently.

**Comment by JordanMarr** [comment]
That seems like a good idea.

**Comment by kgday** [comment]
I have created a PR.

**Comment by JordanMarr** [comment]
Looks good! Released as v1.2.1.

**Reply by kgday** [reply]
Thank you.
